### PR TITLE
Add QueryBuilder vs DQL section

### PR DESCRIPTION
**Pull request by bocharsky-bw**
Describe difference between QueryBuilder and DQL usage.

| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | none |

### CODE REVIEW

File: book/doctrine.rst [hunk 1]
```diff
@@ -793,6 +793,16 @@ For more information on Doctrine's Query Builder, consult Doctrine's
 
 .. _book-doctrine-custom-repository-classes:
```
Comment (xabbuh) [inline book/doctrine.rst]: The label must be moved (it belongs to the following section).

File: book/doctrine.rst [hunk 2]
```diff
@@ -793,6 +793,16 @@ For more information on Doctrine's Query Builder, consult Doctrine's
...
 
+Query Builder vs DQL
+~~~~~~~~~~~~~~~~~~~~
+
+The ``QueryBuilder`` rather to use instead of ``DQL`` when query logic based
+on some result of custom conditions. It allows you to avoid a query string
+concatenating manually. The ``QueryBuilder`` do this concatenation by itself
+while you calling suitable methods. Also, with ``QueryBuilder`` you could easy
+take away some parts of query to separate methods and reuse it in other queries.
+In other cases you probably should use a ``DQL`` syntax to build your queries.
```
Comment (xabbuh) [inline book/doctrine.rst]: I suggest to reword the paragraph to something like this:

> When the actual query depends on dynamic conditions, your code soon becomes hard to read as you start to concatenate strings. In these cases, it's better to use the Doctrine QueryBuilder which provides a fluent interface to dynamically add every part of the query.

What do you think?

Comment (bocharsky-bw) [inline book/doctrine.rst]: I think it much better, thansk!

File: book/doctrine.rst (hunk 1 above)
Comment (bocharsky-bw) [inline book/doctrine.rst]: I moved it below

**Comment by xabbuh** [comment]
:+1:

**Comment by bocharsky-bw** [comment]
Thanks for help with text tweaking!

**Comment by WouterJ** [comment]
Nice! It's always good to explain when to use which option.

However, the section isn't really a versus anymore, it just explains the use-cases for the QueryBuilder. What about merging it with the first paragraph of the QB section?

```diff
-Instead of writing a DQL string, you can alternatively use a helpful object called
-the ``QueryBuilder`` to build that string for you::
+Instead of writing a DQL string, you can alternatively use a helpful object called
+the ``QueryBuilder`` to build that string for you. This is usefull when the actual
+query depends on dynamic conditions, as your code soon becomes hard to read
+with DQL as you start to concatenate strings.
+
+.. code-block:: php
```

Ideally, we would show an example that shows dynamic query building, but that hasn't to be changed in this PR.

**Comment by xabbuh** [comment]
Ah indeed, I missed the fact that we have that paragraph. The suggested change really fits better.

**Comment by bocharsky-bw** [comment]
@WouterJ I applied your comment

File: book/doctrine.rst [hunk 3]
```diff
@@ -766,8 +766,10 @@ covered later), group, etc. For more information, see the official
 Querying for Objects Using Doctrine's Query Builder
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Instead of writing a DQL string, you can alternatively use a helpful object called
-the ``QueryBuilder`` to build that string for you::
+Instead of writing a DQL string, you can use a helpful object called the
+``QueryBuilder`` to build that string for you. This is usefull when the actual query
```
Comment (xabbuh) [inline book/doctrine.rst]: useful

File: book/doctrine.rst [hunk 4]
```diff
@@ -791,6 +793,14 @@ normal ``Query`` object, which can be used to get the result of the query.
 For more information on Doctrine's Query Builder, consult Doctrine's
 `Query Builder`_ documentation.
 
+DQL vs Query Builder
+~~~~~~~~~~~~~~~~~~~~
+
+When the actual query depends on dynamic conditions, your code soon becomes hard
+to read as you start to concatenate strings. In these cases, it's better to use
+the Doctrine QueryBuilder which provides a fluent interface to dynamically add
+every part of the query.
```
Comment (xabbuh) [inline book/doctrine.rst]: I am not sure if we need this anymore with the addition to the other paragraph.

File: book/doctrine.rst (hunk 3 above)
Comment (bocharsky-bw) [inline book/doctrine.rst]: Fixed

File: book/doctrine.rst (hunk 4 above)
Comment (bocharsky-bw) [inline book/doctrine.rst]: I removed it

**Comment by xabbuh** [comment]
:+1:

**Comment by WouterJ** [comment]
:+1:

**Comment by xabbuh** [comment]
Thank you Victor.

**Comment by xabbuh** [comment]
@bocharsky-bw FYI, I merged your pull request into the `2.3` branch. That's why you see it as closed.

**Comment by bocharsky-bw** [comment]
@xabbuh Thanks for information. Will it merge to other branches too?

**Comment by WouterJ** [comment]
Yes, we merge 2.3 into the newer branches from time to time

**Comment by bocharsky-bw** [comment]
Ok, then I think I could delete this branch patch

**Pull request closed without merging**